### PR TITLE
[JENKINS-72107] Introduce `Loadable` interface

### PR DESCRIPTION
**Pull request by Vlatombe**
CloudBees CI uses `load()` method to reload selectively parts of the Jenkins object model from disk, without reloading everything.

This formalizes the interface that gets used for this purpose, even if most of the `Saveable` implementations already had a `load()` method by convention.

I could not find any use case in the community that could rely on this mecanism. For example, configuration as code uses data binding from casc configuration to live objects, but doesn't need to interact with disk.

See [JENKINS-72107](https://issues.jenkins.io/browse/JENKINS-72107).

### Proposed changelog entries

- Developer: Formalize an interface for objects that can be loaded from disk.

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@mention

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

File: core/src/main/java/hudson/model/PersistentDescriptor.java
Comment (jglick) [inline core/src/main/java/hudson/model/PersistentDescriptor.java:16]:
```suggestion
    @PostConstruct
    @Override
```

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (jglick) [inline core/src/main/java/jenkins/model/Jenkins.java:3374]: Why not throw it up? The interface allows it to be. Then internal calls to `loadConfig` could simply be replaced with `load`.

**Review by jglick — APPROVED** [review]

**Review by jtnord — APPROVED** [review]

File: core/src/main/java/hudson/Plugin.java
Comment (jtnord) [inline core/src/main/java/hudson/Plugin.java:270]: as this is opening up the `load` could it be the case that a plugin has made a previous assumption that this would only be called once at Jenkins startup.  Also that it may not be thread safe?

Should this be `synchronized` or have you checked for classes that overload this, or made those synchronized where they need to be (there may be none, given this is deprecated)

Comment (Vlatombe) [reply core/src/main/java/hudson/Plugin.java:270]: I could find the following OSS overloads.

* https://github.com/jenkinsci/global-build-stats-plugin/blob/9867e364d6719cc1e01fe974d3499e29733cf138/src/main/java/hudson/plugins/global_build_stats/GlobalBuildStatsPlugin.java#L47 makes it public
* https://github.com/jenkinsci/docker-traceability-plugin/blob/9d197cc6da0e641e78944441f14ce022f244f550/docker-traceability-plugin/src/main/java/org/jenkinsci/plugins/docker/traceability/DockerTraceabilityPlugin.java#L84-L90 does seem like useless
* https://github.com/jenkinsci/jira-issues-versioning-plugin/blob/73d4a878fcb0bf62c1fd7f75a59dd64f23432be8/hudson-plugin/src/main/java/org/jvnet/hudson/plugins/jira/issueversioning/plugin/hudson/PluginImpl.java#L14 useless

There doesn't seem to be an intensive usage of it (thankfully). Make it `synchronized` as well as `save()` could be considered (same as `Descriptor`) though there seems to be a general point to synchronizing configuration change whether in `configure(..)` `load()` or `save()`

**Comment by NotMyFault** [comment]
/label ready-for-merge

---

This PR is now ready for merge. We will merge it after ~24 hours if there is no negative feedback.
Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process.
Thanks!

**Pull request merged**
